engine/src/main/scala/cromwell/engine/workflow/WorkflowDockerLookupActor.scala

```diff
@@ -0,0 +1,322 @@
+        goto(Running) using newData
+
+      case Failure(e) =>
+        fail(new Exception("Failed to parse docker tag -> hash mappings from DB", e))
+    }
+  }
+
+  private def requestDockerHash(request: DockerHashRequest, data: WorkflowDockerLookupActorData): State = {
+    sendDockerCommand(request)
+    val replyTo = sender()
+    val updatedData = data.copy(hashRequests = data.hashRequests + (request -> List(replyTo)))
+    stay using updatedData
+  }
+
+  private def recordMappingAndRespond(response: DockerHashSuccessResponse, data: WorkflowDockerLookupActorData): State = {
+    // Add the new label to hash mapping to the current set of mappings.
+    val request = response.request
+    // Handle `hashRequests` carefully.  There's a possibility that we could have received a `DockerHashActorTimeout` message
+    // which cleared out all the hash requests in the state data, but then we got a hash response for a request that wasn't
+    // the request which generated the timeout.  The `JobPreparationActor`s corresponding to this request will already have
+    // received lookup failure messages.  tl;dr we shouldn't assume that there's an entry in `hashRequests` corresponding to
+    // the request in this response.
+    val replyTos = data.hashRequests.get(request).toList.flatten
+    replyTos foreach { _ ! DockerHashSuccessResponse(response.dockerHash, request) }
+    val updatedData = data.copy(hashRequests = data.hashRequests - request, mappings = data.mappings + (request.dockerImageID -> response.dockerHash))
+    stay using updatedData
+  }
+
+  private def respondToAllRequests(reason: Throwable,
+                                   data: WorkflowDockerLookupActorData,
+                                   messageBuilder: (Throwable, DockerHashRequest) => WorkflowDockerLookupResponse): WorkflowDockerLookupActorData = {
+    data.hashRequests foreach { case (request, replyTos) =>
+      replyTos foreach { _ ! messageBuilder(reason, request) }
+    }
+    data.clearHashRequests
+  }
+
+  private def respondToAllRequestsWithLookupFailure(reason: Throwable, data: WorkflowDockerLookupActorData): WorkflowDockerLookupActorData = {
+    respondToAllRequests(reason, data, WorkflowDockerLookupFailure.apply)
+  }
+
+  private def respondToAllRequestsWithTerminalFailure(reason: Throwable, data: WorkflowDockerLookupActorData): WorkflowDockerLookupActorData = {
+    respondToAllRequests(reason, data, WorkflowDockerTerminalFailure.apply)
+  }
+
+  private def persistDockerHash(response: DockerHashSuccessResponse, data: WorkflowDockerLookupActorData): State = {
+    val dockerHashStoreEntry = DockerHashStoreEntry(workflowId.toString, response.request.dockerImageID.fullName, response.dockerHash.algorithmAndHash)
+    databaseInterface.addDockerHashStoreEntry(dockerHashStoreEntry) onComplete {
+      case Success(_) => self ! DockerHashStoreSuccess(response)
+      case Failure(ex) => self ! DockerHashStoreFailure(response.request, ex)
+    }
```

Comment:
My comment then was that what we really had was multiple actors-worth of stuff instead of a callback soup to self. Data should generally be moving forward through actors, not looping back within the same one